core/src/main/java/org/apache/iceberg/expressions/ExpressionParser.java

```diff
@@ -250,13 +295,32 @@ private void term(Term term) throws IOException {
       throw new UnsupportedOperationException("Cannot write unsupported term: " + term);
     }
 
-    private void transform(String transform, String name) throws IOException {
+    private void transform(String transform, String name, Integer fieldId) throws IOException {
       gen.writeStartObject();
       gen.writeStringField(TYPE, TRANSFORM);
       gen.writeStringField(TRANSFORM, transform);
-      gen.writeStringField(TERM, name);
+      gen.writeFieldName(TERM);
+      if (fieldId != null) {
+        reference(name, fieldId);
+      } else {
+        gen.writeString(name);
+      }
       gen.writeEndObject();
     }
+
+    private void reference(String name, Integer fieldId) throws IOException {
+      if (fieldId != null) {
+        gen.writeStartObject();
+        gen.writeStringField(TYPE, REFERENCE);
+        if (name != null) {
+          gen.writeStringField(TERM, name);
```

Comment:
outside field name is also called `TERM` at line 302. seems a bit weird to also use `TERM` here.